llvm/lib/Target/AMDGPU/AMDGPUPromoteAlloca.cpp

```diff
@@ -644,6 +645,42 @@ static Value *promoteAllocaUserToVector(Instruction *Inst, const DataLayout &DL,
       auto *SubVecTy = FixedVectorType::get(VecEltTy, NumLoadedElts);
       assert(DL.getTypeStoreSize(SubVecTy) == DL.getTypeStoreSize(AccessTy));
 
+      // If idx is dynamic, then sandwich load with bitcasts.
+      // ie. CurValTy                 SubVecTy  AccessTy
+      //     <64 x i8> ->             <16 x i8> <8 x i16>
+      //     <64 x i8> -> <4 x i128> -> i128 -> <8 x i16>
+      // Extracting subvector with dynamic index has very large expansion in
+      // the amdgpu backend. Limit to pow2 for UDiv.
+      FixedVectorType *AccessVecTy = cast<FixedVectorType>(AccessTy);
+      FixedVectorType *VectorTy = AA.Vector.Ty;
+      uint64_t NumBits =
+          SubVecTy->getScalarSizeInBits() * SubVecTy->getNumElements();
+      uint64_t LoadAlign = cast<LoadInst>(Inst)->getAlign().value();
+      if (!isa<ConstantInt>(Index) && SubVecTy->isIntOrIntVectorTy() &&
+          llvm::isPowerOf2_32(SubVecTy->getNumElements()) &&
+          VectorTy->getNumElements() % SubVecTy->getNumElements() == 0 &&
+          llvm::isPowerOf2_32(AccessVecTy->getNumElements()) &&
+          NumBits <= (LoadAlign * 8u)) {
+        IntegerType *NewElemType = Builder.getIntNTy(NumBits);
+        const unsigned NewNumElts = VectorTy->getNumElements() *
+                                    VectorTy->getScalarSizeInBits() /
+                                    NewElemType->getScalarSizeInBits();
+        const unsigned IndexDivisor = VectorTy->getNumElements() / NewNumElts;
+        assert(VectorTy->getScalarSizeInBits() <
+                   NewElemType->getScalarSizeInBits() &&
+               "New element type should be bigger");
```

Comment:
I only just noticed that. What justifies this assertion? And is it really necessary?

What if we have a `<4 x i16>` load from a `<32 x i32>` alloca?

Reply:
This was just a mental check to say element type of bitcast should always be larger one than alloca & subvector's element type. I had meant to compare SubVecTy's element type but it should be same as VectorTy's element type. I can use SubVecTy's.